Add ReconstructionLoss

akhandait opened the pull request:
splitting #1433
@sumedhghaisas

akhandait commented:
@zoq I have added ann_test_tools.hpp and removed the functions from ann_layer_test.cpp. Please confirm if this is okay.

akhandait commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss_impl.hpp:
```diff
@@ -0,0 +1,62 @@
+{
+  dist = new DistType(input);
+  return dist->LogProbability(target);
+}
```
I think I messed up here, the target should be the one we form the DistType with. The input is the observation to the distribution.

akhandait commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss_impl.hpp:
```diff
@@ -0,0 +1,62 @@
+    OutputType&& output)
+{
+  dist->LogProbBackward(target, output);
+}
```
Same here, instead of `target`, it should be `input`.

zoq commented:
> @zoq I have added ann_test_tools.hpp and removed the functions from ann_layer_test.cpp. Please confirm if this is okay.

Looks good, thanks!

sumedhghaisas commented:
The changes look good to me. Although could you rebase this PR on NormalDistribution one rather than adding normal distribution in the same commit? Will be able to see clear diff about the changes added in this commit. Sorry for being so pedantic. :)

akhandait commented:
I rebased it on normal distribution PR. Everything seems to be working now. I will create a local branch which I will rebase on this one and start working on the model.

akhandait commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp:
```diff
@@ -0,0 +1,89 @@
+
+ private:
+  //! Locally-stored distribution object.
+  DistType* dist;
```
Sorry if this doubt seems silly, but how would I use the constructor in reconstruction_loss_impl.hpp if I already declare 
```
DistType dist;
```
 It will just use the empty default constructor which I don't want to use.

akhandait replied:
Maybe this was forgotten, can you please take a look at this?

rcurtin replied:
Sorry, I hadn't had a chance to get around to it yet (but I should have, sorry about that).  You can just use the constructor initialization list to do this:

```
ReconstructionLoss::ReconstructionLoss() :
    dist(/* arguments to constructor */)
{
  ...
}
```

For the other code where you run `dist = new DistType(std::move(input));` you can just do `dist = DistType(std::move(input));`.

akhandait replied:
Oh sorry, that was really silly. I don’t know why I got confused over that.

rcurtin replied:
No worries, happy to help :)

zoq commented on src/mlpack/tests/ann_dist_test.cpp:
```diff
@@ -0,0 +1,157 @@
+  static const constexpr double log2pi = 1.83787706640934533908193770912475883;
+  arma::mat param, target, gradient;
+  double output;
+  param.ones(10, 1);
```
Just a note you can also write ``arma::mat param = arma::ones(10, 1)``.

akhandait replied:
Just a doubt, is this faster than `param.ones(10, 1);`?

zoq replied:
It probably saves the copy constructor, but that's neglectable here.

zoq commented on src/mlpack/methods/ann/dists/normal_distribution_impl.hpp:
```diff
@@ -0,0 +1,106 @@
+  if (applySoftplus)
+    SoftplusFunction::Fn(preStdDev, stdDev);
+  else
+    stdDev = preStdDev;
```
I wonder if it makes sense to use an advanced constructor here, something like:

```c++
stdDev = arma::mat(preStdDev.memptr(), preStdDev.n_rows, preStdDev.n_cols, false, false);
```

I guess, since we already check ``applySoftplus`` in almost every case we can also use ``preStdDev`` instead of ``stdDev``.

akhandait replied:
Do you mean we should remove `stdDev` completely?

zoq replied:
I'm not sure that's possible because if ``applySoftplus`` is true, we need the output of ``preStdDev``, what I proposed was that we could avoid the copy ``stdDev = preStdDev;`` if ``applySoftplus`` is false.

akhandait replied:
Yeah, it won't be possible to remove `stdDev`. I will use the advanced constructor to save the copy.

zoq commented on src/mlpack/methods/ann/dists/normal_distribution.hpp:
```diff
@@ -0,0 +1,146 @@
+   */
+  double Probability(const DataType&& observation) const
+  {
+    return exp(LogProbability(observation));
```
It might be usefull to use ``std::exp`` here, to avoid any confusion with ``arma::exp``.

akhandait replied:
Okay, will change this.

akhandait commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss_impl.hpp:
```diff
@@ -0,0 +1,62 @@
+    Archive& /* ar */,
+    const unsigned int /* version */)
+{
+  // Nothing to do here.
```
Here `dist` is doing nothing but creating a distribution from the `input` for the forward and backward passes using the `target`. It will change on every forward pass. I don't think it needs to be serialized.

rcurtin replied:
Oh, I see---you are right, sorry about that. :)

akhandait commented:
@zoq I corrected the `Gradient()` function in the Sequential layer but didn't change the `Backward()` function in ffn_impl.hpp. For now, I have added a note that tells users to add an `IdentityLayer<>` before the Sequential layer if it's the first layer of the network.

zoq commented:
@akhandait Sounds reasonable to me, thanks for looking into the issue.

mlpack-jenkins commented:
Can one of the admins verify this patch?

mlpack-jenkins commented:
Can one of the admins verify this patch?

rcurtin commented:
@mlpack-jenkins test this please

zoq commented:
@akhandait what do you think is this ready?

akhandait commented:
@zoq I added one final commit, now it's ready!

zoq commented on src/mlpack/methods/ann/dists/normal_distribution_impl.hpp:
```diff
@@ -0,0 +1,107 @@
+/**
+ * @file normal_distribution.cpp
```
This should be ``normal_distribution_impl.hpp``.

zoq commented on src/mlpack/methods/ann/dists/normal_distribution_impl.hpp:
```diff
@@ -0,0 +1,107 @@
+{
+  if (observation.size() != mean.size())
+  {
+    Log::Fatal << "NormalDistribution<>::NormalDistribution(): The size of the"
```
Missing space at the end or at the beginning of the sentence.

zoq commented on src/mlpack/methods/ann/layer/sequential_impl.hpp:
```diff
@@ -138,15 +138,20 @@ void Sequential<InputDataType, OutputDataType, CustomLayers...>::Gradient(
 {
-  boost::apply_visitor(GradientVisitor(std::move(input), std::move(error)),
-      network.front());
+  boost::apply_visitor(GradientVisitor(std::move(boost::apply_visitor(
```
Nice catch!

zoq commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp:
```diff
@@ -0,0 +1,89 @@
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The mean squared error performance function measures the network's
```
hm, probably not updated?

zoq commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp:
```diff
@@ -0,0 +1,89 @@
+   * Computes the mean squared error function.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
```
Looks like this should be called, ``target``.

zoq commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp:
```diff
@@ -0,0 +1,89 @@
+  ReconstructionLoss();
+
+  /*
+   * Computes the mean squared error function.
```
Looks like this should be updated as well.

zoq commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp:
```diff
@@ -0,0 +1,89 @@
+   */
+  template<typename InputType, typename TargetType>
+  double Forward(const InputType&& input, const TargetType&& target);
+  /**
```
Can you put a new line between the two function here?

zoq commented on src/mlpack/methods/ann/loss_functions/reconstruction_loss_impl.hpp:
```diff
@@ -0,0 +1,63 @@
+    OutputType&& output)
+{
+  dist.LogProbBackward(std::move(target), std::move(output));
+  output = -output;
```
I guess the compiler could optimize this away, but perhaps we could use ``output *= -1`` instead?

akhandait commented:
I will make this changes by tonight. Sorry for so many mistakes.

zoq commented:
No worries, all of them are minor.

akhandait commented on src/mlpack/methods/ann/dists/normal_distribution.hpp:
```diff
@@ -0,0 +1,147 @@
+  /**
+   * Create multiple Normal distributions with the given parameters.
+   *
+   * @param mean The DataType of means of the multiple distributions
```
Yes, I was confused a little here, but I am now sure that matrix would be more appropriate.

akhandait commented on src/mlpack/methods/ann/dists/normal_distribution.hpp:
```diff
@@ -0,0 +1,147 @@
+
+  //! Calculate standard deviation from the pre standard deviation.
+  //! Only meant for testing.
+  void ApplySoftplus() { SoftplusFunction::Fn(preStdDev, stdDev); }
```
Oh! Yeah, it isn't needed.

sumedhghaisas commented:
@zoq Atharva and I are actually working on making VAE work with this loss. Currently we have a good working implementation of VAE that works with MeanSquaredError loss but ReconstructionLoss with NormalDistribution is giving some mathematical problems. Will it be better to merge this when we have a working proof of this loss?

zoq commented:
@sumedhghaisas Ah, I missed that; I agree let's wait until the last issues are solved.

sumedhghaisas commented:
@zoq Actually Atharva and I are communicating through Hangouts because I am at ICML. Thats why you missed it. :) Also I have a question for you. In a VAE setting, P(x) = P(x | z) * p(z). Which distribution is used for modelling P(x | z) when MeanSquaredError is used? Cause VAE_loss = NLL + KL, where NLL will be the likelihood of X in q(X | z) [decoder]. This likelihood is given by MeanSquaredError, but I am unable to understand what distribution it uses? When we model q(x | x) with ormalDistribution we are getting weird results.

sumedhghaisas commented on src/mlpack/methods/ann/dists/bernoulli_distribution.hpp:
```diff
@@ -0,0 +1,132 @@
+   * If the param matrix is already between [0, 1], then applyLogistic can be
+   * set to false.
+   *
+   * @param param The matrix of probabilities or pre probabilities of
```
I think 'logits' is more understandable and common than 'pre probabilities'. Also we should mention the how the dimensions of param affect the distribution.

sumedhghaisas commented on src/mlpack/methods/ann/dists/bernoulli_distribution.hpp:
```diff
@@ -0,0 +1,132 @@
+   *
+   * @param param The matrix of probabilities or pre probabilities of
+   *        the multiple distributions.
+   * @param applyLogsitic If true, then we apply Logistic function to the
```
'If true, we apply', and add a stop in the end of the comment.

sumedhghaisas commented on src/mlpack/methods/ann/dists/bernoulli_distribution.hpp:
```diff
@@ -0,0 +1,132 @@
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Multiple Bernoulli distributions.
```
Here we should rather mention how the distribution is defined. For example, each entry in the param is individual Bernoulli distribution, thus we create num_rows * num_column distributions.Used in ANN framework to define distribution for each feature in each batch, where number of features becomes num_rows and batch size becomes num_columns.

sumedhghaisas commented on src/mlpack/methods/ann/dists/bernoulli_distribution.hpp:
```diff
@@ -0,0 +1,132 @@
+   * Return a matrix of randomly generated samples according to the
+   * probability distributions defined by this object.
+   *
+   * @return Matrix(integer) of random samples from the multiple Bernoulli distributions.
```
We should mention the dimensions of the matrix that is returned and how it relates to params.

sumedhghaisas commented on src/mlpack/methods/ann/dists/bernoulli_distribution.hpp:
```diff
@@ -0,0 +1,132 @@
+  //! Return a modifiable copy of the probability matrix.
+  DataType& Probability() { return probability; }
+
+  //! Return the pre probability matrix.
```
logits

sumedhghaisas commented on src/mlpack/methods/ann/dists/bernoulli_distribution.hpp:
```diff
@@ -0,0 +1,132 @@
+  //! Return the pre probability matrix.
+  const DataType& PreProbability() const { return preProbability; }
+
+  //! Return a modifiable copy of the pre probability matrix.
```
logits

akhandait commented:
@sumedhghaisas @zoq I did some cherry-picking and removed the normal distribution commits(they still are in the NormalDistribution PR). I also created a different PR for BernoulliDistribution and newer changes. This PR only has commits which were reviewed. I think it's ready to merge now.

sumedhghaisas commented:
Ahh I see. We can review Bernoulli Distribution PR and get it merged fairly quickly as its not much of a code. We can merge this just after that. :)
